VolumeManager and Volume internals cleanup

ctubbsii opened the pull request:
* VolumeManager
  * remove getDefaultVolume, which was only used to create a temporary
    directory for creating tables with split points, and to resolve
    relative paths
  * replace "default volume" concept with the first available volume
    (this is what it was effectively doing anyway... but now it's
    implemented more cleanly)
  * simplified much of the volume management code in the CreateTable
    FaTE operation classes, where creation of temporary files for
    storing splits on a pre-split table was being handled
  * rename getFileSystem methods throughout that return VolumeManager to
    getVolumeManager, and stop passing it as an extra parameter if it is
    already available via a ServerContext object
* Volume
  * Remove unneeded overloaded Volume.prefixChild(Path)
  * Rename Volume.isValidPath(Path) to Volume.containsPath(Path) to
    improve readability
  * Update javadoc for containsPath to describe what it means for a Path
    to be contained
  * Update VolumeImplTest to verify behavior of containsPath
* VolumeImpl
  * Make protected fields private
  * Rename conf to hadoopConf to clarify which config it is
  * normalize base path by stripping out trailing slashes
  * rename helper method equivalentPaths(Path) to isAncestorPathOf(Path)
    for readability in containsPath(Path) implementation
  * add more strict argument checking for prefixChild(String)
  * fix bug in isAncestorPathOf(Path) that incorrectly concluded /a/path
    is an ancestor of /a/pa (/a is an ancestor, but /a/pa is not)
* VolumeImplTest
  * check normalization by testing with trailing slashes
  * add more test cases for isAncestorPathOf, including checking for
    "breakout" path terms, like ".."
  * add test cases for prefixChild and containsPath
* Tablet
  * Fix bug in Tablet that wasn't using the Path's FileSystem when
    deleting a file, but instead was using an arbitrary FileSystem

milleruntime commented on core/src/main/java/org/apache/accumulo/core/volume/Volume.java:
```diff
@@ -38,26 +38,25 @@
+   * supplied path should not include any scheme (such as <code>file:</code> or <code>hdfs:</code>),
+   * and should not contain any relative path "breakout" patterns, such as <code>../</code>. If the
+   * path begins with a single slash, it will be preserved while prefixing this volume. If it does
+   * not begin with a single slash, one will be inserted.
```
It looks like you are just clarifying what should be in the string parameter and did not change the behavior correct?

ctubbsii replied:
The only behavior change is the stricter checking of params. Previously, it wouldn't work right... if you passed in a param that had a leading slash, it would not prefix it correctly.

milleruntime commented on server/base/src/main/java/org/apache/accumulo/server/fs/VolumeManager.java:
```diff
@@ -179,12 +179,14 @@ void bulkRename(Map<Path,Path> oldToNewPathMap, int poolSize, String poolName,
-  Volume getDefaultVolume();
+  default Volume getFirst() {
+    return getVolumes().iterator().next();
+  }
```
Nice, this is much more clear.

milleruntime commented on server/manager/src/main/java/org/apache/accumulo/master/FateServiceHandler.java:
```diff
@@ -779,65 +781,37 @@ private String validateNewNamespaceArgument(ByteBuffer namespaceArg, TableOperat
-   * Always check for and delete the splits file if it exists to prevent issues in case of server
-   * failure and/or FateServiceHandler retries.
-   */
-  private void removeAndCreateTempFile(String path) throws IOException {
```
Any idea why this temp file was created? Sounds like it was created to help with failover somewhere.

milleruntime commented on server/manager/src/main/java/org/apache/accumulo/master/tableOps/Utils.java:
```diff
@@ -198,19 +197,21 @@ public static void checkNamespaceDoesNotExist(ServerContext context, String name
    * Given an input stream and a flag indicating if the file info is base64 encoded or not, retrieve
    * the data from a file on the file system. It is assumed that the file is textual and not binary
    * data.
+   *
```
Looks like the comments are outdated now that you changed the parameters.

ctubbsii replied:
Oh, probably. I can check.

ctubbsii replied:
Actually, I think this javadoc is still okay.

milleruntime commented:
I think a comment in the javadoc Property about the first volume essentially being chosen as the "default" volume would be good for users.

ctubbsii commented:
The following ITs failed: TabletIT,SuspendedTabletsIT,ShellServerIT,SimpleBalancerFairnessIT,RegexGroupBalanceIT,MetadataMaxFilesIT,LargeRowIT,CreateInitialSplitsIT,ConfigurableCompactionIT

I'll track these down and fix... probably tomorrow.

milleruntime commented:
Looks good, assuming you can get the ITs to pass.

ctubbsii commented:
I will probably fix the tests next week. I wasn't able to get to them today.

Manno15 commented:
I pulled down these changes to test #1739 against and I am running into a bug trying to create a table with accumulo-testing. I have not taken a deep look into it as of yet but I will try to get to the bottom of it.
The error is: 
~~~
Exception in thread "main" org.apache.accumulo.core.client.AccumuloException: Internal error processing executeFateOperation
	at org.apache.accumulo.core.clientImpl.TableOperationsImpl.doFateOperation(TableOperationsImpl.java:403)
	at org.apache.accumulo.core.clientImpl.TableOperationsImpl.doFateOperation(TableOperationsImpl.java:355)
	at org.apache.accumulo.core.clientImpl.TableOperationsImpl.doTableFateOperation(TableOperationsImpl.java:1694)
	at org.apache.accumulo.core.clientImpl.TableOperationsImpl.create(TableOperationsImpl.java:246)
	at org.apache.accumulo.testing.continuous.CreateTable.main(CreateTable.java:75)
Caused by: org.apache.thrift.TApplicationException: Internal error processing executeFateOperation
	at org.apache.thrift.TServiceClient.receiveBase(TServiceClient.java:79)
	at org.apache.accumulo.core.master.thrift.FateService$Client.recv_executeFateOperation(FateService.java:126)
	at org.apache.accumulo.core.master.thrift.FateService$Client.executeFateOperation(FateService.java:107)
	at org.apache.accumulo.core.clientImpl.TableOperationsImpl.executeFateOperation(TableOperationsImpl.java:282)
	at org.apache.accumulo.core.clientImpl.TableOperationsImpl.doFateOperation(TableOperationsImpl.java:366)

~~~

ctubbsii commented:
> I pulled down these changes to test #1739 against and I am running into a bug trying to create a table with accumulo-testing. The second error message seems to be the cause.

Thanks for taking a look. I think this is the same error that's affecting the broken tests above. I will be working on fixing this today.

ctubbsii commented:
I believe I have fixed the bug. It was a trivial mistake of the type `var y = f(x); use x` should have been `var y = f(x); use y`. However, I took the opportunity to add some more test cases, and I fixed a problem with `VolumeImpl.toString()`

ctubbsii commented:
All ITs pass now. Merging.
